### PR TITLE
jackson codec for Either, Validated, and Ior

### LINKED ISSUE #40: ["Request"] Add a Ktor Client feature to support Either
Is it possible to add a feature for Ktor client similar to the Retrofit call adapter?

### LINKED ISSUE #45: ["Request"] Basic data types for Jackson
What version are you currently using?

What would you like to see?

It's been several times that I've been trying to use the Either data type in my code. Contrary to options, it's something completely missing from the kotlin stlib so it's a real good gateway drug to Arrow. Each time, I had to revert adding Arrow as a dependency to my codebase due to a lack of jackson support.   

When I look into the codebase of arrow-integrations, the only implementation currently present is for NEL and Option. This is which really limited and not sufficient at all.

So please add at least the following types:
* Either
* Validated
* Ior

### PR DESCRIPTION
**Pull request by myuwono**
### in this PR
I've added configurable jackson serialization/deserialization support for:
- `Either<L, R>` 
- `Validated<E, A>`
- `Ior<L, R>`

I needed to also create a couple of reusable generic machineries to create other product types and union types. Those are being put inside an `internal` package currently as I don't intend that to be used by users outside of the libraries yet. These includes:
- `ProductTypeDeserializer<T>` and `ProductTypeSerializer<T>`, and
- `UnionTypeDeserializer<T>` and `UnionTypeSerializer<T>`

in theory we can now add support for any new sealed data types or any tupled types.

I know this is a rather sizeable PR, happy to split this into smaller chunks if that helps with the review.

fixes #45 
fixes #40

### CODE REVIEW

File: README.md
Comment (raulraja) [inline README.md:59]: These are great additions and build on a previous example. The issue here is the last model example is not great. Can we rethink these docs to include a more meaningful model that does not use `Foo` etc.? After all, this would land in public docs. Thanks!

**Review by raulraja — APPROVED** [review]
thanks @myuwono !

File: README.md
Comment (myuwono) [reply README.md:59]: I agree, i think it's probably a better idea if we use some real data structure as an example. Something like this for instance.
```kotlin
data class ArrowUser(
  val id: String,
  val emails: NonEmptyList<EmailAddress>,
  val website: Option<String>
)
```

Illustrating `Either`, `Ior` and `Validated` would be a little strange. I guess we can simply mention that they are supported.

I'll try to come up with something.

Comment (myuwono) [reply README.md]: I've updated the example in readme to reflect that.
```kotlin
val mapper = ObjectMapper()
  .registerKotlinModule()
  .registerArrowModule()
  .setSerializationInclusion(JsonInclude.Include.NON_ABSENT) // will not serialize None as nulls

data class Organization(val name: String, val address: Option<String>, val websiteUrl: Option<URI>)
data class ArrowUser(val name: String, val emails: NonEmptyList<String>, val organization: Option<Organization>)

val arrowKt = Organization("arrow-kt", none(), URI("https://arrow-kt.io").some())
val arrowUser = ArrowUser(
  "John Doe",
  nonEmptyListOf("john@email.com", "john.doe@email.com.au"),
  arrowKt.some()
)

val prettyPrinter = mapper.writerWithDefaultPrettyPrinter()
prettyPrinter.writeValueAsString(user)
// {
//   "name" : "John Doe",
//   "emails" : [ "john@email.com", "john.doe@email.com.au" ],
//   "organization" : {
//     "name" : "arrow-kt",
//     "websiteUrl" : "https://arrow-kt.io"
//   }
// }

val validOrganization: Validated<Nel<String>, Organization> = arrowKt.valid()
prettyPrinter.writeValueAsString(validOrganization)
// {
//   "valid" : {
//     "name" : "arrow-kt",
//     "websiteUrl" : "https://arrow-kt.io"
//   }
// }
```

Comment (myuwono) [reply README.md]: Also added the documentation on how to register it in Spring Boot

File: arrow-integrations-jackson-module/src/main/kotlin/arrow/integrations/jackson/module/internal/InjectField.kt
Comment (nomisRev) [inline arrow-integrations-jackson-module/src/main/kotlin/arrow/integrations/jackson/module/internal/InjectField.kt:3]: Nit: Should this be a static nested class of `ProductTypeDeserializer` to give it better namespacing?

File: arrow-integrations-jackson-module/src/main/kotlin/arrow/integrations/jackson/module/internal/ProjectField.kt
Comment (nomisRev) [inline arrow-integrations-jackson-module/src/main/kotlin/arrow/integrations/jackson/module/internal/ProjectField.kt:5]: Nit: Should this be a static nested class of `UnionTypeDeserializer ` to give it better namespacing? I had to look around a bit to see where this belonged to and why.

**Review by nomisRev — APPROVED** [review]
Looks great!! Thank you so much @myuwono! And sorry for the wait on the review!

File: arrow-integrations-jackson-module/src/main/kotlin/arrow/integrations/jackson/module/internal/InjectField.kt
Comment (myuwono) [reply arrow-integrations-jackson-module/src/main/kotlin/arrow/integrations/jackson/module/internal/InjectField.kt:3]: hmm this is used in both union and product type deserializers, i might need to restructure this a bit. It may make sense to actually use dedicated types for each deserializers. I'll address this

Comment (myuwono) [reply arrow-integrations-jackson-module/src/main/kotlin/arrow/integrations/jackson/module/internal/InjectField.kt]: awesome i've addressed this. We now use each types namespaced under each serializer / deserializer

**Pull request merged**
